sources/imdsclient/src/lib.rs

```diff
@@ -221,92 +246,110 @@ impl ImdsClient {
-                .send()
-                .await
-                .context(error::Request {
-                    method: "GET",
-                    uri: &uri,
-                })?;
-            trace!("IMDS response: {:?}", &response);
-
-            match response.status() {
-                code @ StatusCode::OK => {
-                    info!("Received {}", target.as_ref());
-                    let response_body = response
-                        .bytes()
-                        .await
-                        .context(error::ResponseBody {
-                            method: "GET",
-                            uri: &uri,
-                            code,
-                        })?
-                        .to_vec();
+        timeout(
+            self.retry_timeout,
+            Retry::spawn(retry_strategy(), || async {
+                let session_token = match self.read_token().await? {
+                    Some(session_token) => session_token,
+                    None => self.write_token().await?,
+                };
+                let response = self
+                    .client
+                    .get(&uri)
+                    .header("X-aws-ec2-metadata-token", session_token)
+                    .send()
+                    .await
+                    .context(error::Request {
+                        method: "GET",
+                        uri: &uri,
+                    })?;
+                trace!("IMDS response: {:?}", &response);
+
+                match response.status() {
+                    code @ StatusCode::OK => {
+                        info!("Received {}", target.as_ref());
+                        let response_body = response
+                            .bytes()
+                            .await
+                            .context(error::ResponseBody {
+                                method: "GET",
+                                uri: &uri,
+                                code,
+                            })?
+                            .to_vec();
+
+                        let response_str = printable_string(&response_body);
+                        trace!("Response: {:?}", response_str);
+
+                        Ok(Some(response_body))
+                    }
 
-                    let response_str = printable_string(&response_body);
-                    trace!("Response: {:?}", response_str);
+                    // IMDS returns 404 if no user data is given, or if IMDS is disabled.
+                    StatusCode::NOT_FOUND => Ok(None),
 
-                    return Ok(Some(response_body));
-                }
+                    // IMDS returns 401 if the session token is expired or invalid.
+                    StatusCode::UNAUTHORIZED => {
+                        warn!("Session token is invalid or expired");
+                        self.write_token().await?;
+                        error::TokenRefreshed.fail()
+                    }
```

Comment:
This feels a bit like using exceptions for control flow, but it seems clear enough once it clicks that returning an error here is what prompts Retry to try again.

The logic might be clearer if written like this:
```suggestion
                    // IMDS returns 401 if the session token is expired or invalid.
                    StatusCode::UNAUTHORIZED => {
                        warn!("Session token is invalid or expired");
                        self.clear_token(); // swaps the Option back to None
                        error::TokenInvalid.fail()
                    }
```

Otherwise, we might successfully refresh the token, and then immediately return an error showing that we failed to refresh the token.

This also avoids some conceptual duplication with the token refresh that happens at the top of each iteration.